En documentation (#21)
yiisoft/test-support: docs/en/README.md

# Test Support package

> Some components of the codebase, during their working process, rely on services that meet the requirements of the [PSR-interfaces](https://www.php-fig.org/psr/).

Reviewer: ```suggestion
Some components of the application codebase rely on services implementing [PSR-interfaces](https://www.php-fig.org/psr/).
```

> In order to test such components, the developer often has to write his own stripped or extended realizations 
> of similar PSR implementations in test environment.

Reviewer: ```suggestion
To test such components, the developer often has to write his own tests-sepcific versions of similar PSR implementations.
```

> Test Support package provide ready-to-use implementations of some PSR interfaces, intended solely to help test your code.

Reviewer: ```suggestion
Test Support package provides ready-to-use implementations of some PSR interfaces, intended solely to help testing your code.
```

## Installation

> The preferred way to install this package is through [composer](https://getcomposer.org/download/):

Reviewer 2: ```suggestion
The preferred way to install this package is through [Composer](https://getcomposer.org/download/):
```

```bash
composer require yiisoft/test-support --dev
```

## Container Interface [PSR-11](https://github.com/php-fig/container)

The `SimpleContainer` class is offered as an implementation of `ContainerInterface`. 

> This is simple dependency container, the configuration for which you pass in the constructor.

Reviewer 2: ```suggestion
This is simple dependency container, the definitions configuration for which you pass in the constructor.
```

> The `SimpleContainer` class is offered as an implementation of `ContainerInterface`. 
> 
> This is simple dependency container, the definitions configuration for which you pass in the constructor.
> Despite the simplicity, container is flexible: 2nd parameter of the constructor accepts a Closure.
> This Closure will be called to get 'default' value, if the requested value is not found if the configuration:

Reviewer: ```suggestion
The `SimpleContainer` is a simple dependency container accepting the definitions configuration as the constructor argument. Despite the simplicity, container is flexible: 2nd parameter of the constructor accepts a Closure.
This Closure will be called to get "default" value if the requested value is not found if the configuration:
```

```php
use Yiisoft\Test\Support\Container\SimpleContainer;

$container = new SimpleContainer(
    ['foo' => 'Foo'],
    fn (string $id) => $id === 'bar' ? 'Bar' : 'Not found'
);
$foo = $container->get('foo'); // Foo
$foo = $container->get('bar'); // Bar
$baz = $container->get('baz'); // Not found
```

## Event Dispatcher [PSR-14](https://github.com/php-fig/event-dispatcher)

The `SimpleEventDispatcher` class is offered as an implementation of `EventDispatcherInterface`.

You can pass any number of closure-listeners to the `SimpleEventDispatcher` constructor. 
Listeners should decide, should they process the event or not, as `SimpleEventDispatcher` does not contain complex logic 
for matching an event to a listener.

> simultaneously, the dispatcher is PSR-compliant and consider `StoppableEventInterface`.

Reviewer 2: ```suggestion
Simultaneously, the dispatcher is PSR-compliant and consider `StoppableEventInterface`.
```

> The `SimpleEventDispatcher` class is offered as an implementation of `EventDispatcherInterface`.
> 
> You can pass any number of closure-listeners to the `SimpleEventDispatcher` constructor. 
> Listeners should decide, should they process the event or not, as `SimpleEventDispatcher` does not contain complex logic 
> for matching an event to a listener.
> Simultaneously, the dispatcher is PSR-compliant and consider `StoppableEventInterface`.

Reviewer: ```suggestion
The `SimpleEventDispatcher` is a test-specific event dispatcher. You can pass any number of closure-listeners to its  constructor. SimpleEventDispatcher` does not contain any complex logic for matching an event to a listener. Every listener should decide by itself if it processes an event or not. The dispatcher is PSR-compliant and works with `StoppableEventInterface`.
```

> If your code sent as event to `SimpleEventDispatcher`, then in tests you can check this using the following methods:

Reviewer 2: ```suggestion
If your code sent an event to `SimpleEventDispatcher`, then in tests you can check this using the following methods:
```

> If your code sent an event to `SimpleEventDispatcher`, then in tests you can check this using the following methods:

Reviewer: ```suggestion
You can test for events sent to `SimpleEventDispatcher` using the following methods:
```

- `isObjectTriggered::isObjectTriggered(object $event, int $times = null)` — Exact `$event` object was triggered.

> - `isClassTriggered::isClassTriggered(string $class, int $times = null)` — Event of `$class` class was triggered .

Reviewer 2: ```suggestion
- `isClassTriggered::isClassTriggered(string $class, int $times = null)` — Event of `$class` class was triggered.
```

- `isInstanceOfTriggered::isInstanceOfTriggered(string $class, int $times = null)` — Event with `$class` interface was triggered.

2nd parameter `$times`, is responsible for checking how many times the event has been called.

## Simple Cache [PSR-16](https://github.com/php-fig/simple-cache)

The `SimpleCacheInterface` interface is implemented by two classes:

1. `MemorySimpleCache`, which works similarly to `ArrayCache` from a 
[yiisoft/cache](https://github.com/yiisoft/cache) package, and

2. `SimpleCacheActionLogger`. It stores all commands, sent to `SimpleCacheInterface`.

### MemorySimpleCache

> `MemorySimpleCache`  Does not use external storage, to store cachec values.

Reviewer 2: ```suggestion
The `MemorySimpleCache` class does not use external storage, to store cached values.
```

> The `MemorySimpleCache` class does not use external storage, to store cached values.

Reviewer: ```suggestion
The `MemorySimpleCache` class does not use external storage to store cached values.
```

> Values stores in the array in the object itself and will be destroyed along with the object.
> Use `MemorySimpleCache` in the simple cases, when you dont need to keep track the history of calls to the cache. 

Reviewer: ```suggestion
Values are stored in the array property of the object itself and will be destroyed along with the object.
Use `MemorySimpleCache` in the simple cases when you do not need to keep track of the history of cache access. 
```

> For caching errors simulation, in the public properties `returnOnSet`, `returnOnDelete` and `returnOnClear`
> you can define return values for the corresponding methods of the `SimpleCacheInterface`.

Reviewer: ```suggestion
You can simulate cache errors by setting public properties `returnOnSet`, `returnOnDelete` and `returnOnClear`. These define values returned by the corresponding methods of the `SimpleCacheInterface`.
```

 
### SimpleCacheActionLogger

> The `SimpleCacheActionLogger` class is the decorator for `SimpleCacheInterface`, intermediary, that remembers all passed 
> commands, even if they are invalid (the key contains invalid characters, or is not string at all, for example). 

Reviewer: ```suggestion
The `SimpleCacheActionLogger` class is a decorator for `SimpleCacheInterface`. It remembers all cache calls even if they are invalid such as when the key contains invalid characters, or is not a string at all. 
```

> Use decorator when it's not enough to test the cache state before and after code execution.
> For example, when the testing service supposed to resend the value to the cache after the first failed attempt.  

Reviewer: ```suggestion
Use the decorator when it is not enough to test the cache state before and after code execution.
For example, when the testing service is supposed to resend the value to the cache after the first failed attempt.
```

```php
use Yiisoft\Test\Support\SimpleCache;

$cache = new SimpleCache\MemorySimpleCache();

> //MemorySimpleCache::set() method will be returning false, which corresponds to an error, accourding to PSR

Reviewer: ```suggestion
//MemorySimpleCache::set() method will return false, which is an error according to PSR
```
